### PR TITLE
colmap: init at 3.5

### PR DESCRIPTION
**Pull request by lebastr**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @

### CODE REVIEW

File: pkgs/applications/science/misc/colmap/default.nix [hunk 1]
```diff
@@ -0,0 +1,33 @@
+{ stdenv, lib, fetchFromGitHub, cmake, boost, ceres-solver, eigen,
+  freeimage, glog, libGLU, glew, qt5,
+  cudaSupport ? false, cudatoolkit ? null }:
+
+assert !cudaSupport || cudatoolkit != null;
+
+let boost_static = boost.override {enableStatic = true; };
```
Comment (alexarice) [inline pkgs/applications/science/misc/colmap/default.nix]: Can you have consistent spacing here?

File: pkgs/applications/science/misc/colmap/default.nix [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+in 
+stdenv.mkDerivation rec {
+  version = "3.5";
+  name = "colmap-" + version;
```
Comment (alexarice) [inline pkgs/applications/science/misc/colmap/default.nix]:
```suggestion
  pname = "colmap";
```

File: pkgs/applications/science/misc/colmap/default.nix [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+     owner = "colmap";
+     repo = "colmap";
+     rev = version;
+     sha256 = "1vnb62p0y2bnga173wmjs0lnyqdjikv0fkcxjzxm8187khk2lly8";
+  };
+  
+  buildInputs = [ boost_static ceres-solver eigen freeimage
+                  glog libGLU glew
+		  qt5.qtbase ] ++
+		  lib.optionals cudaSupport [ cudatoolkit ];
```
Comment (alexarice) [inline pkgs/applications/science/misc/colmap/default.nix]:
```suggestion
		  lib.optional cudaSupport cudatoolkit;
```

File: pkgs/applications/science/misc/colmap/default.nix [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  version = "3.5";
+  name = "colmap-" + version;
+  src = fetchFromGitHub {
+     owner = "colmap";
+     repo = "colmap";
+     rev = version;
+     sha256 = "1vnb62p0y2bnga173wmjs0lnyqdjikv0fkcxjzxm8187khk2lly8";
+  };
+  
+  buildInputs = [ boost_static ceres-solver eigen freeimage
```
Comment (alexarice) [inline pkgs/applications/science/misc/colmap/default.nix]: There must be a nicer way to format this list

File: pkgs/applications/science/misc/colmap/default.nix [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+                  glog libGLU glew
+		  qt5.qtbase ] ++
+		  lib.optionals cudaSupport [ cudatoolkit ];
+
+  nativeBuildInputs = [ cmake ];
+
+  meta = with stdenv.lib; {
+    description = "COLMAP is a general-purpose Structure-from-Motion (SfM) and Multi-View Stereo (MVS) pipeline with a graphical and command-line interface.";
```
Comment (alexarice) [inline pkgs/applications/science/misc/colmap/default.nix]: descriptions should be kept to one line. If necessary you can also add a long description

File: pkgs/top-level/all-packages.nix [hunk 6]
```diff
@@ -7371,6 +7371,9 @@ in
 
   colm = callPackage ../development/compilers/colm { };
 
+  colmap = callPackage ../applications/science/misc/colmap { };
+  colmapWithCuda = colmap.override { cudaSupport = true; };
```
Comment (alexarice) [inline pkgs/top-level/all-packages.nix]:
```suggestion
  colmapWithCuda = callPackage ../application/science/misc/colmap { cudaSupport = true; };
```
I'm not actually sure which is preferred so let someone else comment before changing this

File: pkgs/applications/science/misc/colmap/default.nix [hunk 7]
```diff
@@ -0,0 +1,33 @@
+{ stdenv, lib, fetchFromGitHub, cmake, boost, ceres-solver, eigen,
```
Comment (alexarice) [inline pkgs/applications/science/misc/colmap/default.nix]:
```suggestion
{ mkDerivation, lib, fetchFromGitHub, cmake, boost, ceres-solver, eigen,
```

File: pkgs/applications/science/misc/colmap/default.nix [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  freeimage, glog, libGLU, glew, qt5,
+  cudaSupport ? false, cudatoolkit ? null }:
+
+assert !cudaSupport || cudatoolkit != null;
+
+let boost_static = boost.override {enableStatic = true; };
+in 
+stdenv.mkDerivation rec {
```
Comment (alexarice) [inline pkgs/applications/science/misc/colmap/default.nix]:
```suggestion
mkDerivation rec {
```

File: pkgs/applications/science/misc/colmap/default.nix [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  };
+  
+  buildInputs = [ boost_static ceres-solver eigen freeimage
+                  glog libGLU glew
+		  qt5.qtbase ] ++
+		  lib.optionals cudaSupport [ cudatoolkit ];
+
+  nativeBuildInputs = [ cmake ];
+
+  meta = with stdenv.lib; {
```
Comment (alexarice) [inline pkgs/applications/science/misc/colmap/default.nix]:
```suggestion
  meta = with lib; {
```

File: pkgs/top-level/all-packages.nix [hunk 10]
```diff
@@ -7371,6 +7371,9 @@ in
 
   colm = callPackage ../development/compilers/colm { };
 
+  colmap = callPackage ../applications/science/misc/colmap { };
```
Comment (alexarice) [inline pkgs/top-level/all-packages.nix]:
```suggestion
  colmap = libsForQt5.callPackage ../applications/science/misc/colmap { };
```

**Review by alexarice — CHANGES_REQUESTED** [review]
Qt apps need to use qt's mkDerivation instead of `stdenv.mkDerivation`. Otherwise you get errors like 
```
qt.qpa.plugin: Could not find the Qt platform plugin "xcb" in ""
```

File: pkgs/applications/science/misc/colmap/default.nix (hunk 4 above)
Comment (alexarice) [reply pkgs/applications/science/misc/colmap/default.nix]: I would personally put a newline after `[` and before `]` but I get this is personal preference

**Review by alexarice — APPROVED** [review]
Tested with nix-review and both binaries run.
It gives a lot of errors of the form
```
qt.glx: qglx_findConfig: Failed to finding matching FBConfig (8 8 8 0)
```
I don't know if these are a problem. Might be more to do with setup of qt apps on my laptop

**Comment by lebastr** [comment]
I have no such errors. How do you run a colmap? I have tested 
1. colmapWithCuda on the server without gui using command: 
`colmap automatic_reconstructor --image_path ./images/ --workspace_path ./`
2. on my laptop:
`colmap gui`

**Comment by alexarice** [comment]
I just ran `colmap gui`. Honestly I think qt is set up really weirdly on my system because I have never managed to get any theme to work

File: pkgs/top-level/all-packages.nix (hunk 6 above)
Comment (disassembler) [reply pkgs/top-level/all-packages.nix]: if it's short, putting it on a single line is fine. if it's longer typically use:
```
pkgs.callPackages ../path/to {
  foo = true;
  bar = false;
};

**Comment by disassembler** [comment]
@GrahamcOfBorg build colmap colmapWithCuda

**Pull request merged**